contracts/core/EntryPoint.sol

```diff
@@ -141,6 +151,28 @@ contract EntryPoint is IEntryPoint, StakeManager {
         _compensate(beneficiary, collected);
     }
 
+    /**
+     * perform estimation of enture UserOperation.
+     * this method will always revert. it performs full validation of the UserOperation,
+     * but passing NO_VALIDATION_AGGREGATOR to the account.validateUserOperation, to signal it
+     * to ignore signature failure.
+     * This way, this method can be used to estimate entire execution without requiring the user
+     * to sign the request.
+     * Note that in order to collect the the success/failure of the target call, it must be executed
+     * with trace enabled to track the emitted events.
+     */
+    function estimateExecution(UserOperation calldata op) external {
+
+        UserOpInfo memory opInfo;
+
+        _validatePrepayment(SIMULATION_INDEX, op, opInfo, address(0));
```

Comment:
Why do we need to use SIMULATION_INDEX in estimateExecution?  It is used to trigger the code that finds the aggregator instead of calling the aggregator in handleAggregatedOps.  It is not used for skipping the signature check (instead, we just don't call validateDeadline here).  estimateExecution doesn't care about aggregators, so what's the reason for not using zero here?

Reply:
SIMULATION_INDEX has several purposes:
-  instead of receiving an aggregator, we query the wallet for its own aggregator
- it add a NUMBER marker
- revert on proper FailOp for missing contracts for wallet/paymaster

Reply:
And any of them is relevant during estimateExecution?  I think they're all relevant to the bundler (during validation check) but not to a user during wallet-simulation.  If there's value in it, we can leave it here.  I'm just wondering if there is.

Reply:
yes, they are.
- we can't call validateUserOp with wrong aggregator (it reverts), so we must first query for it.
- we do want proper exception strings instead if revert("")
- this is the core of estimateUserOperation, so you do need the marker to determine the boundary for gas usage (actually, it is possible also without it, by full tracking of method signatures and return values)
-

Reply:
Got it.  I thought all these things are only done in validation simulation (validateUserOp) but not when doing full execution simulation (estimateExecution).  Do we also use the marker when simulating the entire op?

Reply:
added marker so that if we will make a traceCall on the simulateExecution, it will still not garble the results.